corehq/messaging/scheduling/models/abstract.py

```diff
@@ -0,0 +1,289 @@
+import pytz
+from corehq.apps.casegroups.models import CommCareCaseGroup
+from corehq.apps.groups.models import Group
+from corehq.apps.locations.dbaccessors import get_all_users_by_location
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.users.models import CommCareUser, WebUser
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from corehq.messaging.scheduling.dbaccessors import save_schedule_instance
+from corehq.util.mixin import UUIDGeneratorMixin
+from corehq.util.timezones.utils import get_timezone_for_domain, coerce_timezone_value
+from datetime import datetime, tzinfo
+from dimagi.utils.decorators.memoized import memoized
+from django.db import models
+from django.core.exceptions import ValidationError
+
+
+class ScheduleForeignKeyMixin(models.Model):
+    timed_schedule_id = models.IntegerField(null=True)
+    alert_schedule_id = models.IntegerField(null=True)
+
+    class Meta:
+        abstract = True
+
+    class NoAvailableSchedule(Exception):
+        pass
+
+    class UnknownScheduleType(Exception):
+        pass
+
+    @property
+    def schedule(self):
+        from corehq.messaging.scheduling.models import TimedSchedule, AlertSchedule
+
+        if self.timed_schedule_id:
+            return TimedSchedule.objects.get(pk=self.timed_schedule_id)
+        elif self.alert_schedule_id:
+            return AlertSchedule.objects.get(pk=self.alert_schedule_id)
+
+        raise self.NoAvailableSchedule()
+
+    @property
+    @memoized
+    def memoized_schedule(self):
+        """
+        This is named with a memoized_ prefix to be clear that it should only be used
+        when the schedule is not changing.
+        """
+        return self.schedule
+
+    @schedule.setter
+    def schedule(self, value):
+        from corehq.messaging.scheduling.models import TimedSchedule, AlertSchedule
+
+        self.timed_schedule_id = None
+        self.alert_schedule_id = None
+
+        if isinstance(value, TimedSchedule):
+            self.timed_schedule_id = value.pk
+        elif isinstance(value, AlertSchedule):
+            self.alert_schedule_id = value.pk
+        else:
+            raise self.UnknownScheduleType()
+
+
+class ScheduleInstance(UUIDGeneratorMixin, ScheduleForeignKeyMixin):
```

Comment:
good point, looking at the docs, there's a [cleaner way](https://docs.djangoproject.com/en/1.10/ref/models/fields/#uuidfield) to autogenerate the uuid - will update